src/associations/has-many.js

```diff
@@ -330,32 +330,20 @@ class HasMany extends Association {
   async set(sourceInstance, targetInstances, options) {
     if (targetInstances === null) {
       targetInstances = [];
+      options = options ? { ...options, destroy: true } : { destroy: true };
     } else {
       targetInstances = this.toInstanceArray(targetInstances);
     }
 
     const oldAssociations = await this.get(sourceInstance, { ...options, scope: false, raw: true });
     const promises = [];
-    const obsoleteAssociations = oldAssociations.filter(old => !targetInstances.some(obj => obj[this.target.primaryKeyAttribute] === old[this.target.primaryKeyAttribute]));
+    const obsoleteAssociations = oldAssociations.filter(old => !targetInstances.some(obj => obj[this.target.primaryKeyAttribute] === old[this.target.primaryKeyAttribute]) || old === null);
     const unassociatedObjects = targetInstances.filter(obj => !oldAssociations.some(old => obj[this.target.primaryKeyAttribute] === old[this.target.primaryKeyAttribute]));
     let updateWhere;
     let update;
 
-    if (obsoleteAssociations.length > 0) {
-      update = {};
-      update[this.foreignKey] = null;
-
-      updateWhere = {
-        [this.target.primaryKeyAttribute]: obsoleteAssociations.map(associatedObject => associatedObject[this.target.primaryKeyAttribute]),
-      };
-
-      promises.push(this.target.unscoped().update(
-        update,
-        {
-          ...options,
-          where: updateWhere,
-        },
-      ));
+    if (obsoleteAssociations.length > 0 || targetInstances == null) {
```

Comment:
```suggestion
    if (obsoleteAssociations.length > 0) {
```

targetInstances can't be null